builtin/credential/aws/path_role.go

```diff
@@ -348,14 +400,98 @@ func (b *backend) pathRoleCreateUpdate(
 		roleEntry.BoundIamInstanceProfileARN = boundIamInstanceProfileARNRaw.(string)
 	}
 
-	// Ensure that at least one bound is set on the role
-	switch {
-	case roleEntry.BoundAccountID != "":
-	case roleEntry.BoundAmiID != "":
-	case roleEntry.BoundIamInstanceProfileARN != "":
-	case roleEntry.BoundIamRoleARN != "":
-	default:
+	if boundIamPrincipalARNRaw, ok := data.GetOk("bound_iam_principal_arn"); ok {
+		roleEntry.BoundIamPrincipalARN = boundIamPrincipalARNRaw.(string)
+	}
+
+	if inferRoleTypeRaw, ok := data.GetOk("inferred_entity_type"); ok {
+		roleEntry.RoleInferredType = inferRoleTypeRaw.(string)
+	}
+
+	if inferredAWSRegionRaw, ok := data.GetOk("inferred_aws_region"); ok {
+		roleEntry.InferredAWSRegion = inferredAWSRegionRaw.(string)
+	}
 
+	var allowEc2Auth, allowIamAuth bool
+
+	if allowedAuthTypesRaw, ok := data.GetOk("allowed_auth_types"); ok {
+		roleEntry.AllowedAuthTypes = strutil.ParseDedupAndSortStrings(allowedAuthTypesRaw.(string), ",")
+	} else if req.Operation == logical.CreateOperation {
+		roleEntry.AllowedAuthTypes = strutil.ParseDedupAndSortStrings(data.Get("allowed_auth_types").(string), ",")
+	}
+
```

Comment:
Sorry, I don't understand your question.

`AllowedAuthTypes` is solely about what kind of **authentication** is allowed, i.e., what kind of material is sufficient to establish an identity. If it's `ec2` then the instance identity document; if it's `iam` then a signed `GetCallerIdentity` request.

`RoleInferredType` deals solely with **authorization** when using an `iam` auth type.

So, when `RoleInferredType` is set and `AllowedAuthTypes` doesn't contain `ec2` then the desired behavior is that the only thing that should be allowed to authenticate is an IAM principal using a signed `GetCallerIdentity` request in which the RoleSessionName is a valid EC2 instance ID, and we apply EC2-specific binds to the instance given by the instance ID.

What am I missing here? Does the code not behave as I just described? Is the documentation confusing? Do you disagree with the desired behavior? Something else?